design/custom-request-timeout.md

```diff
@@ -0,0 +1,256 @@
+# Executive Summary
+
+**Status**: _Draft_
+
+This document describes the design of a new resource in IngressRoute for custom request timeout and retries. 
+
+This new resource will be integrated into Contour 0.11.
+
+# Goals
+
+- Generic resource name for configuring custom request timeouts per IngressRoute
+- Generic resource name for configuring custom retry attempts per IngressRoute
+
+# Non-goals
+
+- None
+
+# Background
+
+Contour supports custom request timeout and custom retry attempts via Ingress. The same functionality has been
+requested via IngressRoute as well.
+
+# High-level design
+
+At a high level, this document proposes the changes in the IngressRoute's CRD to also support custom request
+timeouts and retries. It covers:
+
+- Fields added for timeout and retries in the YAML file.
+- A high-level coverage on the struct changes required in the code.
+- Arguments passed for Envoy API calls.
+
+## Proposed YAML fields
+
+```yaml
+apiVersion: contour.heptio.com/v1beta1
+kind: IngressRoute
+metadata:
+    name: github.com
+    namespace: prod
+spec:
+    virtualhost:
+        fqdn: www.github.com
+        routes:
+        - match: /contour
+          # the timeout applied to the requests in this route
+          timeoutPolicy:
+              # the request timeout applied on this route, it is the maximum time that a client will await
+              # a response to its request.
+              request: 500ms
+              # the maximum time that an idle connection is kept open when there are no active requests
+              idle: 2s
+          retryPolicy:
+              # the number of retries it should make on failure
+              maxRetries: "7"
```

Comment:
Nice catch. Will do that.

Reply:
Nice. I like how you separated timeoutPolicy and retryPolicy into separate objects, this feels right.

Reply:
This is the part I don't really get. If I understand it correctly, the `timeoutPolicy` here will be mapped to RouteAction.timeout  and RouteAction.idle_timeout in Envoy API. So making it into its own object, just to make it look cleaner, but not for later to be used or referred from other objects, correct?

Reply:
your understanding about `timeoutPolicy` to RouteAction.timeout and RouteAction.idle_timeout is correct. Within contour, timeout will be in its own object only for cleanliness, while populating to Envoy API it will be mapped accordingly.